File: app/src/main/java/com/jggdevelopment/simpleweather/fragments/SettingsFragment.java
package com.jggdevelopment.simpleweather.fragments;


import android.app.FragmentManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;

import com.jggdevelopment.simpleweather.MainActivity;
import com.jggdevelopment.simpleweather.R;

import androidx.core.app.ActivityCompat;
import androidx.fragment.app.FragmentTransaction;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceScreen;
import androidx.preference.SwitchPreference;
import androidx.preference.SwitchPreferenceCompat;

import static android.Manifest.permission.ACCESS_FINE_LOCATION;

public class SettingsFragment extends PreferenceFragmentCompat {

    private SwitchPreferenceCompat useCurrentLocationPreference;
    private SwitchPreferenceCompat useCelsiusPreference;
    private SwitchPreferenceCompat useDarkThemePreference;
    private SharedPreferences prefs;
    private PreferenceScreen preferenceScreen;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.settings);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        // To get a preference
        preferenceScreen = getPreferenceScreen();

        prefs = getActivity().getSharedPreferences("com.jggdevelopment.simpleweather", Context.MODE_PRIVATE);

        setupPreferenceViews();
        setUpPreferenceListeners();

    }

    public void turnOnLocation() {
        ActivityCompat.requestPermissions(getActivity(), new String[] {ACCESS_FINE_LOCATION}, 200);
    }

    private void setupPreferenceViews() {
        useCurrentLocationPreference = (SwitchPreferenceCompat) preferenceScreen.findPreference("useLocation");
        useCelsiusPreference = (SwitchPreferenceCompat) preferenceScreen.findPreference("useCelsius");
        useDarkThemePreference = (SwitchPreferenceCompat) preferenceScreen.findPreference("useDarkTheme");

        if (prefs.getBoolean("useCurrentLocation", false)) {
            useCurrentLocationPreference.setChecked(true);
        } else {
            useCurrentLocationPreference.setChecked(false);
        }

        if (prefs.getBoolean("useCelsius", false)) {
            useCelsiusPreference.setChecked(true);
        } else {
            useCelsiusPreference.setChecked(false);
        }

        if (prefs.getBoolean("useDarkTheme", false)) {
            useDarkThemePreference.setChecked(true);
        } else {
            useDarkThemePreference.setChecked(false);
        }
    }

    private void setUpPreferenceListeners() {
        useCurrentLocationPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                if (newValue == Boolean.TRUE) {
                    prefs.edit().putBoolean("useCurrentLocation", true).commit();
                } else {
                    prefs.edit().putBoolean("useCurrentLocation", false).commit();
                }
                turnOnLocation();
                return true;
            }
        });

        useCelsiusPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                if (newValue == Boolean.TRUE) {
                    prefs.edit().putBoolean("useCelsius", true).commit();
                } else {
                    prefs.edit().putBoolean("useCelsius", false).commit();
                }
                return true;
            }
        });

        useDarkThemePreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                if (newValue == Boolean.TRUE) {
                    prefs.edit().putBoolean("useDarkTheme", true).commit();
                } else {
                    prefs.edit().putBoolean("useDarkTheme", false).commit();
                }

                getActivity().recreate();

                return true;
            }
        });
    }
}

File: app/src/main/java/com/jggdevelopment/simpleweather/fragments/SettingsFragment.kt
package com.jggdevelopment.simpleweather.fragments

import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import androidx.core.app.ActivityCompat
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceScreen
import androidx.preference.SwitchPreferenceCompat
import com.jggdevelopment.simpleweather.R

class SettingsFragment : PreferenceFragmentCompat() {

    private lateinit var useCurrentLocationPreference: SwitchPreferenceCompat
    private lateinit var useCelsiusPreference: SwitchPreferenceCompat
    private lateinit var useDarkThemePreference: SwitchPreferenceCompat
    private lateinit var prefs: SharedPreferences

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        prefs = activity!!.getSharedPreferences("com.jggdevelopment.simpleweather", Context.MODE_PRIVATE)

        setupPreferenceViews()
        setupPreferenceListeners()
    }

    private fun turnOnLocation() {
        ActivityCompat.requestPermissions(activity!!, arrayOf(ACCESS_FINE_LOCATION), 200)
    }

    private fun setupPreferenceViews() {
        useCurrentLocationPreference = preferenceScreen.findPreference<Preference>("useLocation") as SwitchPreferenceCompat
        useCelsiusPreference = preferenceScreen.findPreference<Preference>("useCelsius") as SwitchPreferenceCompat
        useDarkThemePreference = preferenceScreen.findPreference<Preference>("useDarkTheme") as SwitchPreferenceCompat

        useCurrentLocationPreference.isChecked = prefs.getBoolean("useCurrentLocation", false)
        useCelsiusPreference.isChecked = prefs.getBoolean("useCelsius", false)
        useDarkThemePreference.isChecked = prefs.getBoolean("useDarkTheme", false)
    }

    private fun setupPreferenceListeners() {
        useCurrentLocationPreference.setOnPreferenceChangeListener { preference, newValue ->
            prefs.edit().putBoolean("useCurrentLocation", newValue as Boolean).commit()
        }

        useCelsiusPreference.setOnPreferenceChangeListener { preference, newValue ->
            prefs.edit().putBoolean("useCelsius", newValue as Boolean).commit()
        }

        useDarkThemePreference.setOnPreferenceChangeListener { preference, newValue ->
            prefs.edit().putBoolean("useDarkTheme", newValue as Boolean).commit()
            activity!!.recreate()
            true
        }
    }



    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.settings)
    }
}